src/lib/pubkey/hss_lms/lms.h

```diff
@@ -0,0 +1,342 @@
+                                    LMS_Tree_Node_Idx q,
+                                    const LMS_Message& msg) const;
+
+   private:
+      LMS_Seed m_seed;
+};
+
+class LMS_Signature;
+
+/**
+ * @brief The LMS public key.
+ *
+ * Format according to RFC 8554:
+ * u32str(type) || u32str(otstype) || I || T[1]
+ */
+class BOTAN_TEST_API LMS_PublicKey : public LMS_Instance {
+   public:
+      /**
+       * @brief Parse a public LMS key.
+       *
+       * @param slicer The BufferSlicer at the public key bytes' position
+       * @return The LMS public key.
+       * @throws Decoding_Error If parsing the public key fails.
+       */
+      static LMS_PublicKey from_bytes_or_throw(BufferSlicer& slicer);
+
+      /**
+       * @brief Construct a public key for given public key data
+       */
+      LMS_PublicKey(LMS_Params lms_params, LMOTS_Params lmots_params, LMS_Identifier I, LMS_Tree_Node lms_root);
+
+      /**
+       * @brief Construct a new public key from a given LMS private key (RFC 8554 5.3).
+       */
+      LMS_PublicKey(const LMS_PrivateKey& sk);
+
+      /**
+       * @brief Bytes of the full lms public key according to 8554 5.3
+       *
+       * pub_key_bytes = u32str(type) || u32str(otstype) || I || T[1]
+       */
+      std::vector<uint8_t> to_bytes() const;
+
+      /**
+       * @brief The expected size of an LMS public key for given @p lms_params
+       */
+      static size_t size(const LMS_Params& lms_params);
+
+      /**
+       * @brief Verify a LMS signature.
+       *
+       * See RFC 8554 5.4.2 - Algorithm 6.
+       *
+       * @param msg The signed message.
+       * @param sig The already parsed LMS signature.
+       * @return True if the signature is valid, false otherwise.
+       */
+      bool verify_signature(const LMS_Message& msg, const LMS_Signature& sig) const;
+
+   private:
+      /**
+       * @brief Compute an lms public key candidate.
+       *
+       * Given the LMS public key, a LMS-Signature-LMS_Message pair, compute
+       * an LMS public key candidate as described in RFC 8554 5.4.2 Algorithm 6a.
+       */
+      std::optional<LMS_Tree_Node> lms_compute_root_from_sig(const LMS_Message& msg, const LMS_Signature& sig) const;
+
+      /**
+       * @brief Root node of the LMS tree ('T[1]' in RFC 8554 5.3)
+       */
+      const LMS_Tree_Node& lms_root() const { return m_lms_root; }
+
+      LMS_Tree_Node m_lms_root;
+};
+
+/**
+ * @brief Container for LMS Signature data.
+ *
+ * Contains a method for secure signature parsing.
+ */
+class BOTAN_TEST_API LMS_Signature {
+   public:
+      /**
+       * @brief Parse the bytes of a lms signature into a LMS Signature object
+       *
+       * @param slicer A BufferSlicer object at the position of the LMS_Signature to parse
+       * @return LMS_Signature object
+       * @throws Decoding_Error If parsing the signature fails.
+       */
+      static LMS_Signature from_bytes_or_throw(BufferSlicer& slicer);
```

Comment:
We chose this interface for LMS signatures because it is challenging to precompute the length of the signature sections. We need to parse the first fields of the signature to obtain the used LMS instance. Then, the first fields of the OTS signature define the OTS instance. Only then can we reconstruct the length. 

Fishing those values from the entire signature is messy.  Alternatively, we could pass the entire span as input and return a subspan of the remaining bytes. That's exactly the information the BufferSlicer contains.